### PR TITLE
add rule nodejs.module

### LINKED ISSUE #5938: rule for nodejs addon
### Is your feature request related to a problem? Please describe.

A rule to build nodejs addon.

### Describe the solution you'd like

For meson, it like:

```meson
project('C++ node example addon', 'cpp')

inc_node = include_directories('/usr/include/node', is_system: true)

# For compatibility with node-gyp we use "shared_module" rather than
# "shared_library". The difference important in the current context is that
# shared_module will happily allow undefined references at link-time. It may or may
# not be what you want, feel free to change to `shared_library`, in which case you
# have to also link it with find_library('node')
shared_module(
  'hello',
  ['hello.cc'],
  include_directories : [inc_node],
  name_prefix: '',     # don't prepend "lib"
  name_suffix: 'node', # required for addon to load properly
)

node = find_program('node')

test('addon',
     node,
     args: ['-e',
            'const addon = require("./hello.node");'
            + 'console.log(addon.hello());'
           ]
    )
```

https://stackoverflow.com/questions/58633310/how-to-compile-node-js-native-api-extension-via-meson-build-system

### Describe alternatives you've considered

_No response_

### Additional context

Related: #5125

An example: https://github.com/tonyfettes/coc-rime/blob/master/xmake.lua

### PR DESCRIPTION
**Pull request by Freed-Wu**
Remember to `npm add -D node-addon-api node-api-headers` to install header files
for C++ and C.

example:

```lua
target("foo")
do
    set_languages("cxx17")
    add_rules("nodejs.module")
    add_includedirs("node_modules/node-addon-api", "node_modules/node-api-headers/include")
    add_files("*.cc")
end
```

* Before adding new features and new modules, please go to issues to submit the relevant feature description first.
* Write good commit messages and use the same coding conventions as the rest of the project.
* Please commit code to dev branch and we will merge into master branch in feature
* Ensure your edited codes with four spaces instead of TAB.

------

* 增加新特性和新模块之前，请先到issues提交相关特性说明，经过讨论评估确认后，再进行相应的代码提交，避免做无用工作。
* 编写友好可读的提交信息，并使用与工程代码相同的代码规范，代码请用4个空格字符代替tab缩进。
* 请提交代码到dev分支，如果通过，我们会在特定时间合并到master分支上。
* 为了规范化提交日志的格式，commit消息，不要用中文，请用英文描述。

~Wait https://github.com/xmake-io/xmake-repo/pull/6004~

Fix #5938

### CODE REVIEW

File: xmake/rules/nodejs/module/xmake.lua [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+-- Copyright (C) 2015-present, TBOOX Open Source Group.
+--
+-- @author      ruki
+-- @file        xmake.lua
+--
+
```
Comment (waruqi) [inline xmake/rules/nodejs/module/xmake.lua:20]: usage

File: xmake/rules/nodejs/module/xmake.lua [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        -- export symbols
+        if target:is_plat("windows") then
+            target:add("shflags", "/export:napi_register_module_v1", "/export:node_api_module_get_api_version_v1",
+                { force = true })
+        else
+            target:set("symbols", "none")
+        end
+
+        -- https://github.com/nodejs/node-addon-api/issues/1021
+        if is_plat("mingw") then
```
Comment (waruqi) [inline xmake/rules/nodejs/module/xmake.lua:44]: target:is_plat

File: xmake/rules/nodejs/module/xmake.lua [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            import("core.project.config")
```
Comment (waruqi) [inline xmake/rules/nodejs/module/xmake.lua:45]: move imports to the first line

File: xmake/rules/nodejs/module/xmake.lua [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            local outputdir = path.join(target:autogendir(), "/node-api-stub")
```
Comment (waruqi) [inline xmake/rules/nodejs/module/xmake.lua:46]: remove `/` if use path.join

File: xmake/rules/nodejs/module/xmake.lua [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+            if not os.isdir(outputdir) then
+                import("devel.git")
```
Comment (waruqi) [inline xmake/rules/nodejs/module/xmake.lua:48]: move to the first line

File: xmake/rules/nodejs/module/xmake.lua [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+-- limitations under the License.
+--
+-- Copyright (C) 2015-present, TBOOX Open Source Group.
+--
+-- @author      ruki
+-- @file        xmake.lua
+--
+
+rule("nodejs.module")
+    on_load(function(target)
```
Comment (waruqi) [inline xmake/rules/nodejs/module/xmake.lua:22]: please use `on_config` instead of on_load

File: xmake/rules/nodejs/module/xmake.lua [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+            target:set("symbols", "none")
+        end
+
+        -- https://github.com/nodejs/node-addon-api/issues/1021
+        if is_plat("mingw") then
+            import("core.project.config")
+            local outputdir = path.join(target:autogendir(), "/node-api-stub")
+            if not os.isdir(outputdir) then
+                import("devel.git")
+                git.clone("https://github.com/napi-bindings/node-api-stub", { depth = 1, outputdir = outputdir })
```
Comment (waruqi) [inline xmake/rules/nodejs/module/xmake.lua:49]: we should move it to xmake-repo, and use `add_packages` + target:pkg("xx") to get it.

File: xmake/rules/nodejs/module/xmake.lua [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+        -- add node library
+        local has_node = false
+        local includedirs = get_config("includedirs") -- pass node library from nodejs/xmake.lua
+        if includedirs and includedirs:find("node-api-headers", 1, true) then
+            has_node = true
+        end
+        if not has_node then
+            -- user use `add_requires/add_packages` to add node package
+            for _, pkg in ipairs(target:get("packages")) do
```
Comment (waruqi) [inline xmake/rules/nodejs/module/xmake.lua:67]: no this values. only `target:pkg("xxx"):installdir()`

File: xmake/rules/nodejs/module/xmake.lua [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+                if pkg == "node-api-headers" then
+                    has_node = true
+                    break
+                end
+            end
+        end
+        if not has_node then
+            target:add(find_package("node-api-headers"))
```
Comment (waruqi) [inline xmake/rules/nodejs/module/xmake.lua:75]: please do not use find_package

File: xmake/rules/nodejs/module/xmake.lua [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+        end
+
+        -- set library name
+        local modulename = target:name():split('.', { plain = true })
+        modulename = modulename[#modulename]
+        target:set("filename", modulename .. ".node")
+
+        -- add node library
+        local has_node = false
+        local includedirs = get_config("includedirs") -- pass node library from nodejs/xmake.lua
```
Comment (waruqi) [inline xmake/rules/nodejs/module/xmake.lua:61]: please do not use `get_config`

File: xmake/rules/nodejs/module/xmake.lua [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+                if pkg == "node-api-headers" then
+                    has_node = true
+                    break
+                end
+            end
+        end
+        if not has_node then
+            target:add(find_package("node-api-headers"))
+        end
+    end)
```
Comment (waruqi) [inline xmake/rules/nodejs/module/xmake.lua:77]: newline

File: xmake/rules/nodejs/module/xmake.lua (hunk 7 above)
Comment (Freed-Wu) [reply xmake/rules/nodejs/module/xmake.lua:59]: Refer https://github.com/xmake-io/xmake-repo/pull/6004

**Comment by Freed-Wu** [comment]
Any  other problem?

**Pull request merged**